runtime/compiler/control/CompilationThread.cpp

```diff
@@ -6398,11 +6398,24 @@ TR::CompilationInfoPerThreadBase::preCompilationTasks(J9VMThread * vmThread,
             if (!(flags & J9SHR_AOT_METHOD_FLAG_INVALIDATED))
                {
 #ifdef COMPRESS_AOT_DATA
-               J9JITDataCacheHeader *dataCacheHeader = (J9JITDataCacheHeader*)*aotCachedMethod;
-               TR_AOTMethodHeader *aotMethodHeader = (TR_AOTMethodHeader*)(dataCacheHeader + 1);
+
+
+               TR_AOTMethodHeader *aotMethodHeader = (TR_AOTMethodHeader*)(((J9JITDataCacheHeader*)*aotCachedMethod) + 1);
                if (aotMethodHeader->flags & TR_AOTMethodHeader_CompressedMethodInCache)
                   {
-                  int sizeOfCompressedDataInCache = dataCacheHeader->size;
+                  /**
+                    * For each AOT compiled method we store in the shareclass cache we have follwing layout.
+                    * ---------------------------------------------------------------------------------------
+                    * | CompiledMethodWrapper | J9JITDataCacheHeader | AOTMethodHeader | Metadata | Codedata|
+                    * ---------------------------------------------------------------------------------------
+                    * When we compile and store a method, it stores J9RomMethod for the method in the CompiledMethodWrapper along side
+                    * size of metadata and codedata in the cache. Now when we request a method from shared class cache
+                    * it returns address that points to J9JITDataCacheHeader. As while decompressing the data, we need information about
+                    * the size of original data (We get that information from AOTMethodHeader) and size of deflated data in the cache (Which is stored in
+                    * CompiledMethodWrapper)
+                    * So to access data size in shared class cache, we need to access CompiledMethodWrapper.
+                    */
+                  int sizeOfCompressedDataInCache = *((uint32_t*)(*aotCachedMethod)-2);
```

Comment:
Why did you go back to `*((uint32_t*)(*aotCachedMethod)-2);`?

Reply:
https://github.com/eclipse/openj9/pull/5926#pullrequestreview-253600216